extensions/amp-ad-network-doubleclick-impl/0.1/safeframe-host.js

```diff
@@ -0,0 +1,615 @@
+      expandHeight += Number(this.iframe_.height);
+    }
+    this.handleSizeChange(expandHeight,
+        expandWidth,
+        SERVICE.EXPAND_RESPONSE);
+  }
+
+  /**
+   * @private
+   */
+  handleCollapseRequest_() {
+    this.handleSizeChange(this.initialSize_.height,
+        this.initialSize_.width,
+        SERVICE.COLLAPSE_RESPONSE,
+        /** isCollapse */ true);
+  }
+
+  /**
+   * @param {number} height
+   * @param {number} width
+   */
+  resizeIframe(height, width) {
+    if (this.iframe_) {
+      setStyles(this.iframe_, {
+        'height': height + 'px',
+        'width': width + 'px',
+      });
+    }
+  }
+
+  /**
+   * Resizes the safeframe, and potentially the containing amp-ad element.
+   * Then sends a response message to the Safeframe creative.
+   * @param {number} height In pixels.
+   * @param {number} width In pixels.
+   * @param {string} message
+   * @param {boolean=} optIsCollapse Whether this is a collapse attempt.
+   */
+  handleSizeChange(height, width, message, optIsCollapse) {
+    // If the new size is fully contained within the bounds of the amp-ad,
+    // we can resize immediately as there will be no reflow.
+    if (!optIsCollapse &&
+        width <= this.creativeSize_.width &&
+        height <= this.creativeSize_.height) {
+      this.resizeIframe(height, width);
+      this.sendResizeResponse(/** SUCCESS */ true, message);
+    } else {
+      this.resizeAmpAdAndSafeframe(
+          height, width, message, optIsCollapse);
+    }
+  }
+
+  /**
+   * @param {boolean} success
+   * @param {string} message
+   */
+  sendResizeResponse(success, message) {
+    this.sendMessage_({
+      uid: this.uid,
+      success,
+      newGeometry: this.getCurrentGeometry(),
+      'expand_t': this.currentGeometry_['allowedExpansion_t'],
+      'expand_b': this.currentGeometry_['allowedExpansion_b'],
+      'expand_r': this.currentGeometry_['allowedExpansion_r'],
+      'expand_l': this.currentGeometry_['allowedExpansion_l'],
+      push: true,
+    }, message);
+  }
+
+  /**
+   *
+   * @param {number} height
+   * @param {number} width
+   * @param {string} message
+   * @param {boolean=} optIsCollapse
+   */
+  resizeAmpAdAndSafeframe(height, width, message, optIsCollapse) {
+    this.baseInstance_.attemptChangeSize(height, width).then(() => {
+      const success = !!this.baseInstance_.element.style.height.match(height)
+              && !!this.baseInstance_.element.style.width.match(width);
+      // If the amp-ad element was successfully resized, always update
+      // the size of the safeframe as well. If the amp-ad element could not
+      // be resized, but this is a collapse request, then only collapse
+      // the safeframe.
+      if (success || optIsCollapse) {
+        this.resizeIframe(height, width);
+        this.baseInstance_.element.getResources().resources_.forEach(
+            resource => {
+              if (resource.element == this.baseInstance_.element) {
```

Comment:
Is there a way to wait for the measure as opposed to force it?  These seems ugly to me and ideally the AMP runtime would handle for you.

Reply:
@lannka can you weigh in here on a better way to do this? I agree this is probably not the best way to do this